### PR TITLE
add pause metric in prometheus for scaledobject 

### PR DESCRIPTION
**Pull request by geoffrey1330**
_Provide a description of what has been changed_

### Checklist

- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #

Relates to #4430 
Relates to https://github.com/kedacore/keda-docs/pull/1111

### CODE REVIEW

**Comment by geoffrey1330** [comment]
Hi @zroubalik here is my PR and I made all the fix referenced in #4446

**Comment by geoffrey1330** [comment]
Hi @zroubalik could you please review my PR

File: pkg/metricscollector/prommetrics.go
Comment (JorTurFer) [inline pkg/metricscollector/prommetrics.go:133]: Could you add this metric also to open telemetry?

File: tests/sequential/prometheus_metrics/prometheus_metrics_test.go
Comment (JorTurFer) [inline tests/sequential/prometheus_metrics/prometheus_metrics_test.go:244]: Could you update opentelemetry test too?

File: tests/sequential/opentelemetry_metrics/opentelemetry_metrics_test.go
Comment (JorTurFer) [inline tests/sequential/opentelemetry_metrics/opentelemetry_metrics_test.go:41]: You have to get this metric from current collector url, if you check the KEDA's endpoint, it doesn't check if the metric is sent via otel

Comment (geoffrey1330) [reply tests/sequential/opentelemetry_metrics/opentelemetry_metrics_test.go:41]: How can I get the current collector url

Comment (geoffrey1330) [reply tests/sequential/opentelemetry_metrics/opentelemetry_metrics_test.go]: I have updated the function to use this collector url 
`kedaOperatorCollectorPrometheusExportURL`

File: CHANGELOG.md
Comment (JorTurFer) [inline CHANGELOG.md:63]: I think that it should be an error here, there are 3 extra lines added from previous release, aren't?

**Review by JorTurFer — COMMENTED** [review]
Looking good! Only one stuff with changelog
@zroubalik PTAL too

File: CHANGELOG.md
Comment (geoffrey1330) [reply CHANGELOG.md:63]: Oh so what line should I remove?

Comment (JorTurFer) [reply CHANGELOG.md:63]: I think that only this line is yours: `- **Prometheus Metrics**: Introduce number of paused ScaledObjects in Prometheus metrics ([#4430](https://github.com/kedacore/keda/issues/4430))`, so keep it only

Comment (geoffrey1330) [reply CHANGELOG.md:63]: Alright I will do that

File: CHANGELOG.md
Comment (JorTurFer) [inline CHANGELOG.md:62]: Could you move this to new section?

Comment (geoffrey1330) [reply CHANGELOG.md:62]: new section where?

Comment (geoffrey1330) [reply CHANGELOG.md:62]: Could you please explain more.
I don't get

Comment (JorTurFer) [reply CHANGELOG.md:62]: image

Comment (geoffrey1330) [reply CHANGELOG.md:62]: Oh alright I will do that

Comment (geoffrey1330) [reply CHANGELOG.md]: Done

File: CHANGELOG.md
Comment (JorTurFer) [inline CHANGELOG.md:59]:
```suggestion
```

File: CHANGELOG.md
Comment (JorTurFer) [inline CHANGELOG.md:66]:
```suggestion
- **Prometheus Metrics**: Introduce number of paused ScaledObjects in Prometheus metrics ([#4430](https://github.com/kedacore/keda/issues/4430))
```

File: controllers/keda/scaledobject_controller.go
Comment (JorTurFer) [inline controllers/keda/scaledobject_controller.go:201]: We try (there are still missing places) to print logs in lower letters
```suggestion
		reqLogger.V(1).Info("detected ScaledObject with pause annotation", "targetKind", ref.Kind, "targetName", ref.Name)
```

File: pkg/metricscollector/opentelemetry.go
Comment (JorTurFer) [inline pkg/metricscollector/opentelemetry.go:269]: I know that it's how it's done in other gauges, but I'm not totally sure if we have to create an observable gauge on each execution. 
I have quickly reviewed the SDK but I haven't seen anything yet. Do you know about this @zroubalik ?

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:201]:
```suggestion
		reqLogger.V(1).Info("Detected ScaledObject with pause annotation", "targetKind", ref.Kind, "targetName", ref.Name)
```

Error messages should be in lowercase.

File: pkg/metricscollector/opentelemetry.go
Comment (zroubalik) [inline pkg/metricscollector/opentelemetry.go:271]: don't use Println here

File: pkg/metricscollector/opentelemetry.go
Comment (zroubalik) [reply pkg/metricscollector/opentelemetry.go:269]: I agree, we should try to find a better solution if there's any. Maybe try to ask in OTEL community or try to find suggested examples.

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [reply controllers/keda/scaledobject_controller.go:201]: Also, this shouldn't be here at all. The exisiting code for this logic is below in `econcileScaledObject()`.

**Review by zroubalik — CHANGES_REQUESTED** [review]
Please make sure to only set/unset the Prometheus metrics for paused state in this PR. The functionality for pausing scaling is already there.

The original PR that you used as a source for your work is quite old, in the meantime there were changes in the code and in the way we pause ScaledObjects. For example see this PR: https://github.com/kedacore/keda/pull/4809

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:199]: why is this added here?

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:606]: this is not needed, right? Because we delete the ScaledObject, we don't have to track the Pause state. Or maybe am I missing anything?

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:592]: better to move this to `reconcileScaledObject()` where we already check paused annotations

**Comment by geoffrey1330** [comment]
@zroubalik Could you please review my PR when you have the time.
cc @JorTurFer

File: pkg/metricscollector/opentelemetry.go
Comment (zroubalik) [inline pkg/metricscollector/opentelemetry.go:267]: We shouldn't record "Total number", but mark whether the current ScaledObject is paused - do it the very same way like ScaledObject Activity metrics are done.

File: pkg/metricscollector/prommetrics.go
Comment (zroubalik) [inline pkg/metricscollector/prommetrics.go:130]: Same remark as for opentelemtry - don't track "total number" but just mark if ScaledObject is paused - take an inspiration from Activity

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:207]: this is not needed, take namespace from ScaledObject object

File: pkg/metricscollector/prommetrics.go
Comment (zroubalik) [inline pkg/metricscollector/prommetrics.go:124]: also, please move this code next to the code where we set activity, so we have similar stuff together in the code

File: pkg/metricscollector/prommetrics.go
Comment (zroubalik) [inline pkg/metricscollector/prommetrics.go:253]: again, move this closer to activity

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:213]: here we should only record metrics when we change the status from runing -> paused once and then again once when we set from paused -> running

File: tests/sequential/prometheus_metrics/prometheus_metrics_test.go
Comment (zroubalik) [inline tests/sequential/prometheus_metrics/prometheus_metrics_test.go:857]: test here following:

1. have running ScaledObject -> check that paused metric is set to false
2. pause the SO -> check that  paused metric is set to true
3. unpause the SO -> check that paused metric is set to false

File: tests/sequential/opentelemetry_metrics/opentelemetry_metrics_test.go
Comment (zroubalik) [inline tests/sequential/opentelemetry_metrics/opentelemetry_metrics_test.go:314]: test here following:

1. have running ScaledObject -> check that paused metric is set to false
2. pause the SO -> check that  paused metric is set to true
3. unpause the SO -> check that paused metric is set to false

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [reply controllers/keda/scaledobject_controller.go:213]: set it in the code where we are (un)setting the condition

File: CHANGELOG.md
Comment (zroubalik) [inline CHANGELOG.md:66]: Again, we are not exposing number of paused ScaledObjects

Please be thorough in your PR.

File: controllers/keda/scaledobject_finalizer.go
Comment (zroubalik) [inline controllers/keda/scaledobject_finalizer.go:80]: Don't include not neccessary changes in the PR, these empty lines are not doing anything

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:235]: Just for the sake of clarity, please put this after we set the condition.

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [reply controllers/keda/scaledobject_controller.go:213]: this should go below following line `conditions.SetPausedCondition(metav1.ConditionTrue, kedav1alpha1.ScaledObjectConditionPausedReason, msg)`

**Review by zroubalik — CHANGES_REQUESTED** [review]

**Review by zroubalik — CHANGES_REQUESTED** [review]

**Comment by geoffrey1330** [comment]
@zroubalik could you please review my PR again. as i have fixed all suggested changes. 
cc @JorTurFer

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:231]:
```suggestion
			metricscollector.RecordScalerPaused(scaledObject.Namespace, scaledObject.Name, true)
```

**Review by zroubalik — COMMENTED** [review]
@geoffrey1330 there are still unresolved comments on this PR

**Comment by geoffrey1330** [comment]
> @geoffrey1330 there are still unresolved comments on this PR

I will go ahead and mark all unresolved comments as resolved.

**Comment by geoffrey1330** [comment]
> > > @geoffrey1330 there are still unresolved comments on this PR
> > 
> > 
> > I will go ahead and mark all unresolved comments as resolved.
> 
> No, it means that these things should be resolved.

Resolved most of them. will check for the one's I'm yet to resolve.

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [reply controllers/keda/scaledobject_controller.go:197]: @geoffrey1330 don't Resolve commits, that are not clearly resolved.

I don't understand why are you deleteing this unrelevant line here?

File: pkg/metricscollector/prommetrics.go
Comment (SpiritZhou) [inline pkg/metricscollector/prommetrics.go:191]: Is this function named RecordScaledObjectPaused better as there will be scaledjob paused metric? WDYT @zroubalik

Comment (zroubalik) [reply pkg/metricscollector/prommetrics.go:191]: Good catch, you are right!

File: pkg/metricscollector/opentelemetry.go
Comment (geoffrey1330) [reply pkg/metricscollector/opentelemetry.go:271]: @zroubalik could you please update the code with the new logging format so i can replicate and resolve this conversation.

File: pkg/metricscollector/prommetrics.go
Comment (geoffrey1330) [reply pkg/metricscollector/prommetrics.go:191]: @zroubalik also waiting for your reply on the above question.

File: pkg/metricscollector/opentelemetry.go
Comment (zroubalik) [reply pkg/metricscollector/opentelemetry.go:271]: https://github.com/kedacore/keda/pull/5095

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [reply controllers/keda/scaledobject_controller.go:197]: Okay. But please don't do it in this PR. Each PR should only contain relevant changes.
Better approach is to open a new issue and discuss it there.

Comment (geoffrey1330) [reply controllers/keda/scaledobject_controller.go:197]: Alright. noted

File: pkg/metricscollector/opentelemetry.go
Comment (geoffrey1330) [reply pkg/metricscollector/opentelemetry.go]: @zroubalik I'm yet to figure out the proper way to tackle this issue. 
I think maybe we should create a separate issue for this. WDYT?

Comment (zroubalik) [reply pkg/metricscollector/opentelemetry.go]: okay, that's fair. Please open another issue and provide details about the problem, so we can track it and don't forget about it. Thanks!

Comment (geoffrey1330) [reply pkg/metricscollector/opentelemetry.go]: Alright Let me do that now.

Comment (zroubalik) [reply pkg/metricscollector/opentelemetry.go]: Please put a link to this specific conversation to the issue. This link: https://github.com/kedacore/keda/pull/5045/files/486ca2f1b1dd59d60cfbb86f5b5829264f8cd697#r1348641815

Comment (geoffrey1330) [reply pkg/metricscollector/opentelemetry.go]: Alright

Comment (geoffrey1330) [reply pkg/metricscollector/opentelemetry.go]: Done here is the issue 
https://github.com/kedacore/keda/issues/5101

**Comment by zroubalik** [comment]
> /run-e2e sequential **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/6575137873)

@geoffrey1330 unfortunately the e2e has failed, have you tried to run it locally?

**Comment by geoffrey1330** [comment]
> > /run-e2e sequential **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/6575137873)
> 
> @geoffrey1330 unfortunately the e2e has failed, have you tried to run it locally?

Not yet. Let me try that now.

**Comment by geoffrey1330** [comment]
> > /run-e2e sequential **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/6575137873)
> 
> @geoffrey1330 unfortunately the e2e has failed, have you tried to run it locally?

@zroubalik I'm using ` make e2e-test-local` to test locally any advise on this?

**Comment by geoffrey1330** [comment]
@zroubalik during testing locally, it generated some `*.pb.go` file. should I go ahead and delete this file or rather leave them?

**Comment by zroubalik** [comment]
> @zroubalik during testing locally, it generated some `*.pb.go` file. should I go ahead and delete this file or rather leave them?

Please omit these files from this PR, it is just a change in the version of the tool that generates these files.

**Comment by geoffrey1330** [comment]
> > @zroubalik during testing locally, it generated some `*.pb.go` file. should I go ahead and delete this file or rather leave them?
> 
> Please omit these files from this PR, it is just a change in the version of the tool that generates these files.

Done

**Comment by geoffrey1330** [comment]
Hi @zroubalik could you please trigger /run-e2e sequential

**Comment by geoffrey1330** [comment]
> /run-e2e sequential **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/6624608567)

@tomkerkhove Could you please trigger `/run-e2e sequential ` again

**Comment by geoffrey1330** [comment]
Hi @zroubalik could you please trigger `/run-e2e sequential` as I've fixed all failing test

**Comment by geoffrey1330** [comment]
> /run-e2e sequential **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/6628679559)

@zroubalik I see the test failed but it passed successfully locally.

**Comment by geoffrey1330** [comment]
Hi @zroubalik could you please trigger /run-e2e sequential as I've fixed all failing test

**Comment by geoffrey1330** [comment]
Hi @JorTurFer could you please trigger /run-e2e sequential the initial trigger passed successfully but I fixed DCO issue and would love to test e2e also on this new change

**Comment by geoffrey1330** [comment]
Hi @zroubalik could you please review my PR as all e2e test passed successfully.
cc @JorTurFer @tomkerkhove

File: CHANGELOG.md
Comment (JorTurFer) [inline CHANGELOG.md:70]:
```suggestion
- **Prometheus Metrics**: Introduce paused ScaledObjects in Prometheus metrics ([#4430](https://github.com/kedacore/keda/issues/4430))
```

**Review by JorTurFer — APPROVED** [review]
LGTM!
@SpiritZhou , could you include this metric on your PR improving OTEL process once it's merged?

File: CHANGELOG.md
Comment (JorTurFer) [inline CHANGELOG.md:71]:
```suggestion
- **Prometheus Metrics**: Introduce paused ScaledObjects in Prometheus metrics ([#4430](https://github.com/kedacore/keda/issues/4430))
- **Pulsar Scaler**: support endpointParams in pulsar oauth ([#5069](https://github.com/kedacore/keda/issues/5069))
```

**Review by JorTurFer — APPROVED** [review]

**Review by zroubalik — APPROVED** [review]
LGTM

**Pull request merged**
